packages/core/src/lib/stream_manager.ts

```diff
@@ -0,0 +1,68 @@
+import type { PeerUpdate } from "@libp2p/interface";
+import type { Stream } from "@libp2p/interface/connection";
+import { Peer } from "@libp2p/interface/peer-store";
+import { Libp2p } from "@waku/interfaces";
+import { selectConnection } from "@waku/utils/libp2p";
+import debug from "debug";
+
+export class StreamManager {
+  private streamPool: Map<string, Promise<Stream>>;
+
+  constructor(
+    public multicodec: string,
+    public getConnections: Libp2p["getConnections"],
+    public addEventListener: Libp2p["addEventListener"],
+    private log: debug.Debugger
+  ) {
+    this.addEventListener(
+      "peer:update",
+      this.handlePeerUpdateStreamPool.bind(this)
+    );
+    this.getStream = this.getStream.bind(this);
+    this.streamPool = new Map();
+  }
+
+  public async getStream(peer: Peer): Promise<Stream> {
+    const peerIdStr = peer.id.toString();
+    const streamPromise = this.streamPool.get(peerIdStr);
```

Comment:
Why did you choose this way for implementation? 

Did you try creating streams per peer and storing them in a map and reusing later on? 
I doubt that we actually need to create a stream each time.

Reply:
trying to check if possible for our use case 

but as I see `gossipSub` retains them (though in a wrapper) - [here](https://github.com/ChainSafe/js-libp2p-gossipsub/blob/a2981f956d1f427fae9deb9c769f0baab0db5d26/src/index.ts#L2218)

Reply:
It's just a spec question, ie the protocol spec need to specify if you can reuse a stream or not
Outside of gossipsub, most libp2p native protocol won't read past a single message according to their spec, but you are free to specify in the light push spec that multiple messages should be handled, and implement accordingly in go nim & js

Reply:
added this to appropriate epic  https://github.com/waku-org/js-waku/issues/1521